versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java

```diff
@@ -0,0 +1,416 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.persist.mongodb;
+
+import static org.projectnessie.versioned.persist.serialize.ProtoSerialization.protoToKeyList;
+import static org.projectnessie.versioned.persist.serialize.ProtoSerialization.toProto;
+
+import com.google.protobuf.ByteString;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.mongodb.ConnectionString;
+import com.mongodb.ErrorCategory;
+import com.mongodb.MongoClientSettings;
+import com.mongodb.MongoWriteException;
+import com.mongodb.client.FindIterable;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.UpdateOptions;
+import com.mongodb.client.result.DeleteResult;
+import com.mongodb.client.result.InsertManyResult;
+import com.mongodb.client.result.InsertOneResult;
+import com.mongodb.client.result.UpdateResult;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.bson.Document;
+import org.bson.types.Binary;
+import org.projectnessie.versioned.Hash;
+import org.projectnessie.versioned.ReferenceConflictException;
+import org.projectnessie.versioned.persist.adapter.CommitLogEntry;
+import org.projectnessie.versioned.persist.adapter.KeyList;
+import org.projectnessie.versioned.persist.adapter.KeyListEntity;
+import org.projectnessie.versioned.persist.adapter.KeyWithType;
+import org.projectnessie.versioned.persist.adapter.spi.DatabaseAdapterUtil;
+import org.projectnessie.versioned.persist.nontx.NonTransactionalDatabaseAdapter;
+import org.projectnessie.versioned.persist.nontx.NonTransactionalOperationContext;
+import org.projectnessie.versioned.persist.serialize.AdapterTypes.GlobalStateLogEntry;
+import org.projectnessie.versioned.persist.serialize.AdapterTypes.GlobalStatePointer;
+import org.projectnessie.versioned.persist.serialize.ProtoSerialization;
+
+public class MongoDatabaseAdapter
+    extends NonTransactionalDatabaseAdapter<MongoDatabaseAdapterConfig> {
+
+  private static final String GLOBAL_POINTER = "global_pointer";
+  private static final String GLOBAL_LOG = "global_log";
+  private static final String COMMIT_LOG = "commit_log";
+  private static final String KEY_LIST = "key_list";
+
+  private final ByteString keyPrefix;
+  private final byte[] globalPointerKey;
```

Comment:
Can we make this a `String`?

Reply:
possibly it's because of a security concern, since strings stay in memory, whereas arrays are cleared once code is done with it.

Reply:
We could use `String` of course, but it looks like all the `persist.proto` objects convert to byte arrays naturally (including keys). So I figured we could use a byte array for the global pointer too.

Reply:
It's not about the `byte[]` per se, it's to represent the `globalPointerKey` in a say "mongo db type friedly" way.

Reply:
Reworked to `String` in 1f3f1e4c3afa0f1266f7906e09a58fb549318f66